parcels/compilation/codegenerator.py

```diff
@@ -1014,8 +1014,12 @@ def generate(self, funcname, field_args, const_args, kernel_ast, c_include):
         update_state = c.Assign("particles->state[pnum]", "res")
         update_pdt = c.If("_next_dt_set == 1",
                           c.Block([c.Assign("_next_dt_set", "0"), c.Assign("particles->dt[pnum]", "_next_dt")]))
+        dt_pos = c.If("fabs(endtime - particles->time[pnum])<fabs(particles->dt[pnum])",
+                            c.Block([c.Assign("__dt", "fabs(endtime - particles->time[pnum])"), c.Assign("reset_dt", "1")]),
+                            c.Block([c.Assign("__dt", "fabs(particles->dt[pnum])"), c.Assign("reset_dt", "0")]))
```

Comment:
here also: try the speed-trick from the python version. `abs(...)` becomes `fans(...)` and `math.ceil(...)` becomes `ceil(...)`. `alpha` and `value` may need to be declared here, so this overall becomes a `c.Block([...])`, but that's overall not an issue, I presume.